### PR TITLE
Support for Boolean and Numeric type display options validation

### LINKED ISSUE #386: Implement display options for Number types.
## Problem
The Number Mathesar type supports the following display options:
- Whether to show the data as a percentage.
- The locale to format the number in.

## Solution
(1) We should store these display options in the following format in the `display_options` field of the corresponding column.
```
{
  "show_as_percentage": true,
  "locale": "en_US"
}
```

(2) We should also validate these so that:
- Only columns of this type can have these display options. They should not be able to be set if the column is of a different type.
- `show_as_percentage` should be a boolean
- `locale` should be a valid locale

(3) If the column type is changed, the display options should be deleted.

(4) We should add supported display options to the `types` endpoint.

## Additional Context
- [Design of Number type options on Figma](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?page-id=4260%3A37440&node-id=4270%3A39634&viewport=324%2C48%2C0.29&scaling=contain&starting-point-node-id=4270%3A39634&show-proto-sidebar=1)
- Blocked by #658
- #392

### LINKED ISSUE #390: Implement display options for Boolean types
## Problem
The boolean data type supports the following display options:
- Whether to show the field as a checkbox or a dropdown.
- Whether to show custom labels for `TRUE` and `FALSE`
- Custom label names for `TRUE` and `FALSE`

## Solution
(1) We should store these display options in the following format in the `display_options` field of the corresponding column.
```
{
  "input": "dropdown",
  "use_custom_labels": true,
  "custom_labels": {
    "TRUE": "yes",
    "FALSE": "no"
  }
}
```

(2) We should also validate these so that:
- Only columns of this type can have these display options. They should not be able to be set if the column is of a different type.
- `input` should be set to either `checkbox` or `dropdown`
- `use_custom_labels` should be a boolean
- `custom_labels` can only have the keys `TRUE` and `FALSE`

(3) If the column type is changed, the display options should be deleted.

(4) We should add supported display options to the `types` endpoint.

## Additional Context
- [Design for Boolean type options on Figma](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?page-id=4260%3A37440&node-id=4308%3A43205&viewport=324%2C48%2C0.29&scaling=contain&starting-point-node-id=4270%3A40031&show-proto-sidebar=1)
- Blocked by #658
- #392

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #390, Fixes #386

Adds validating boolean type columns display options. 

**Technical details**

Display options field is a JSON, which would accept any data, so a validation serializer depending upon the type of the column is used to validate the data.

Feature Checklist:
- [X] Store these display options in the following format in the `display_options` field of the corresponding column.

```
{
  "input": "dropdown",
  "use_custom_labels": true,
  "custom_labels": {
    "TRUE": "yes",
    "FALSE": "no"
  }
}
```

- [X] Only columns of this type can have these display options. They should not be able to be set if the column is of a different type.
- [X] input should be set to either checkbox or dropdown
- [X] use_custom_labels should be a boolean
- [X] custom_labels can only have the keys TRUE and FALSE
- [x] If the column type is changed, the display options should be deleted.
- [x] Add supported display options to the types endpoint.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: mathesar/api/display_options.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+                         "enum": ['dropdown', 'checkbox']},
+                        {'name': "custom_labels", "type": "object",
+                         "items": [{"name": "TRUE", "type": "string"},
+                                   {'name': "FALSE", "type": "string"}]}]
+
+        },
+    MathesarTypeIdentifier.NUMBER.value:
+        {
+            "options": [{"name": "show_as_percentage", "type": "boolean"},
+                        {"name": "locale", "type": "string"}]
```
Comment (dmos62) [inline mathesar/api/display_options.py:16]: Should the backend be aware of localization preferences? My instinct is to have the frontend take care of that. So you could have two people working on the same table using different locales.

Comment (silentninja) [reply mathesar/api/display_options.py:16]: Display options are used to store (default) representational preferences that will be used by the frontend, it won't be used by the backend. Please have a look at @kgodey [comment](https://github.com/centerofci/mathesar/issues/392#issuecomment-969168596) that concludes this

Comment (dmos62) [reply mathesar/api/display_options.py:16]: Interesting, I wonder in what scenario a frontend would find backend-defined defaults useful.

Comment (silentninja) [reply mathesar/api/display_options.py:16]: Display options are not defined/generated by the backend, it is set by the user on the frontend using the [`/columns/` API](https://github.com/centerofci/mathesar/blob/f4c2b32d3fd8320f8880a54c951243ae4254f969/mathesar/api/serializers/columns.py#L42). The display options returned as part of `/types/` API is used by the client to know about the available display options for that particular type

**Comment by kgodey** [comment]
@silentninja Please leave this as a draft until #815 is merged since this depends on it. Otherwise, it might accidentally be merged if people aren't paying attention to the base branch.

Also, minor note on the PR description:

> Fixes #390, #386

This will need to say "Fixes #390, fixes #386" in order for both issues to get closed automatically by GitHub.

File: mathesar/api/display_options.py (hunk 1 above)
Comment (dmos62) [reply mathesar/api/display_options.py:16]: I had misunderstood what I was reading. Thanks for bearing with me!

**Review by dmos62 — APPROVED** [review]
Nice work! Did you see any alternatives to the polymorphic serializer?

**Comment by silentninja** [comment]
> Nice work! Did you see any alternatives to the polymorphic serializer?

We can use  conditional validations but this seems like a reusable abstraction, the validations, however, can be a strategy or template pattern based on the use case

**Review by mathemancer — APPROVED** [review]
Looks good to me as well.

**Review by kgodey — APPROVED** [review]
Looks good to me. I have some style issues with lists and tuples, but I think it's fine for now, at some point we should work on a Python style guide and ensure the codebase complies with it.

**Pull request merged**